snap/validate.go

```diff
@@ -727,6 +727,50 @@ func (layout *Layout) constraint() LayoutConstraint {
 	return mountedTree(path)
 }
 
+// layoutRejectionList contains directories that cannot be used as layout
+// targets.  Nothing there, or underneath can be replaced with $SNAP or
+// $SNAP_DATA content, even from the point of view of a single snap.
+var layoutRejectionList = []string{
+	// There's no known reason to allow snaps to replace things there.
+	"/boot",
+	// The /dev directory contains essential device nodes and there's no valid
+	// reason to allow snaps to replace it.
+	"/dev",
+	// The firmware is sometimes loaded on demand by the kernel, in response to
+	// a process performing generic I/O to a specific device. In that case the
+	// mount namespace of the process is searched, by the kernel, for the
+	// firmware. Therefore firmware must not be replicable to prevent malicious
+	// firmware from attacking the host.
+	"/lib/firmware",
+	// Similarly the kernel will load modules and the modules should not be
+	// something that snaps can tamper with.
+	"/lib/modules",
```

Comment:
Absolutely, yes, lets not allow this, but is the reasoning accurate true? Ie, if I try add a firewall rule via iptables, iptables doesn't load the module, but the kernel finds it. I would be surprised that the kernel would look in iptables mount namespace and load something from in there. That would be really bad for untrusted containers. IIRC, we were thinking that we wouldn't let a snap that shipped kmod or something that itself looks in /lib/modules to accidentally use what the snap shipped. I say 'accidentally' cause this is only going to be allowed if the snap plugs kernel-module-control and if it does, it doesn't need to look in /lib/modules, it can just load them. So this entry would assist in protecting a snap that used kernel-module-control from accidentally loading from itself rather than system modules. Please correct me if I missed something.

Reply:
Initially this is how I though this works but now I'm not so sure. Checking.

Reply:
I cannot find any poof either way yet. I'll check one more place and change the comment to include either the reference or remove the comment.

Reply:
We had a look at this and the kernel *does* load modules internally by calling modprobe on the host. There are several tricky aspects of this, that we were not able to trace yet, such as the exact nature of the mount namespace used. A quick experiment seems to suggest it the initial mount namespace but we were unable to pinpoint this exactly in the kernel source code.